payments/utils.py

```diff
@@ -0,0 +1,104 @@
+import calendar
+from datetime import date, timedelta
+from decimal import Decimal, ROUND_HALF_UP
+from functools import wraps
+
+from dateutil.utils import today
+
+from leases.utils import (
+    calculate_berth_lease_end_date,
+    calculate_berth_lease_start_date,
+)
+
+
+def round_price(price):
+    return price.quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
+
+
+def rounded(func):
+    """
+    Decorator for rounding function result
+    """
+
+    @wraps(func)
+    def wrapped(*args, **kwargs):
+        value = func(*args, **kwargs)
+        return round_price(value)
+
+    return wrapped
+
+
+def currency_format(value):
+    return f"{round_price(value)}€"
+
+
+def currency(func):
+    @wraps(func)
+    def wrapped(*args, **kwargs):
+        value = func(*args, **kwargs)
+        return currency_format(value)
+
+    return wrapped
+
+
+def percentage_format(value):
+    return f"{round_price(value)}%"
+
+
+def percentage(func):
+    @wraps(func)
+    def wrapped(*args, **kwargs):
+        value = func(*args, **kwargs)
+        return percentage_format(value)
+
+    return wrapped
+
+
+def convert_aftertax_to_pretax(aftertax_price, tax_percentage):
+    return aftertax_price / (1 + tax_percentage / 100)
+
+
+def calculate_order_due_date(delta=14):
+    return date.today() + timedelta(days=delta)
+
+
+@rounded
+def calculate_product_partial_month_price(
+    base_price: Decimal, start_date: date, end_date: date
+) -> Decimal:
+    _, days_in_month = calendar.monthrange(start_date.year, start_date.month)
+    delta = (end_date - start_date).days
+    price = (delta * base_price) / days_in_month
```

Comment:
How would we bill for every month? The `order` is generated for/with a `lease`, with a determined duration. But how are the clients billed for that? I.e. where would we say what's the price that the customer has to pay for either month (in this case)?

Reply:
nonono, I donät mean billing every months, I mean the total price calculation when you have a product with `period == MONTH`

Reply:
AHHHH, okay, gotcha. I was under the assumption that monthly products would only be assigned to "monthly" products. So, I guess that instead of counting `days_in_month` we can just calculate the day difference.

Reply:
How about the "full month" case? Say:
```
start_date = june 15
end_date = september 22
```

We calculate partially `june` and `september`, but for `july + august` there's two options:
1. Add the full month price _as-is_.
2. Again, calculate the "partial" month price to adjust the base monthly price to either 30-31 days (or 28-29)

Reply:
so:
1. full month price as is for july and for august, since those months are rented from the first day to the last
2. for june you: (take month price / 30 days in june) * 16 days of june that the user is renting it for
3. for september you: (take month price / 30 days in september) * 22 days of september that the user is renting it for
(4). if it would have been october 22 as the end date: (take month price / 31 days in october) * 22 days of october that the user is renting it for